onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -145,6 +145,35 @@ Status GetNCHWInput(ModelBuilder& model_builder, const Node& node, size_t input_
   return Status::OK();
 }
 
+// Transpose layouts if necessary for element wise operators with 2 inputs
+// and return the layout type of output tensor
+// If both inputs have same layout, the output will have the same layout
+// Otherwise we will need transpose the nhwc input back to nchw, and output will be nchw
+Status TransposeBinaryOpInputLayout(ModelBuilder& model_builder, const Node& node,
+                                    size_t input1_idx, size_t input2_idx,
+                                    std::string& input1, std::string& input2,
+                                    bool& output_is_nhwc) ORT_MUST_USE_RESULT;
+Status TransposeBinaryOpInputLayout(ModelBuilder& model_builder, const Node& node,
+                                    size_t input1_idx, size_t input2_idx,
+                                    std::string& input1, std::string& input2,
+                                    bool& output_is_nhwc) {
+  bool input1_is_nhwc = model_builder.IsOperandNHWC(input1);
+  bool input2_is_nhwc = model_builder.IsOperandNHWC(input2);
+  output_is_nhwc = false;
+
+  if (input1_is_nhwc == input2_is_nhwc) {
+    output_is_nhwc = input1_is_nhwc;
+  } else if (input1_is_nhwc) {
+    // need transpose input1 back to nchw
+    ORT_RETURN_IF_ERROR(GetNCHWInput(model_builder, node, input1_idx, input1));
+  } else {  // input2_is_nhwc
+    // need transpose input2 back to nchw
+    ORT_RETURN_IF_ERROR(GetNCHWInput(model_builder, node, input2_idx, input2));
+  }
+
```

Comment:
Just wondering about the setup here. 

Would it be better to convert the nchw input to nhwc to prefer nhwc ouput, so that if the downstream node that consumes the output runs on NNAPI it's input is in nhwc format? Possibly we need something more advanced to know if the consumer of the output will run on NNAPI to know the best approach though. If the consumer is running on NNAPI I assume it's better to transpose one input to nhwc and have output in nhwc (vs. transposing one input to nchw and also having to transpose the output to nhwc if used by a downstream node running on nnapi).

Reply:
For now the layout change is very simple and on the safe side, if you don't know what to do, convert back to nchw since that is the default behavior. For example, elementwise ops, there are other things to consider, such as broadcast, instead if transpose, we will need add reshape to the nchw input.

I agree we can do better to look at the consumer of the output,, or doing something similar to optimizing the entire graph, to reduce possible transpose. I will put note in the task https://msdata.visualstudio.com/Vienna/_queries/edit/923623/?fullScreen=true&triage=true